[JSC] DFGLoopUnrollingPhase: Support partial loop unrolling for non-constant bounds

hyjorc1 opened the pull request:
#### b16849008e438b42d44b16e43da8b99c0796cf87
<pre>
[JSC] DFGLoopUnrollingPhase: Support partial loop unrolling for non-constant bounds
<a href="https://bugs.webkit.org/show_bug.cgi?id=290335">https://bugs.webkit.org/show_bug.cgi?id=290335</a>
<a href="https://rdar.apple.com/147775029">rdar://147775029</a>

Reviewed by Yusuke Suzuki.

This patch enhances the DFG loop unrolling phase to support partial unrolling
for loops with non-constant bounds. Previously, loop unrolling was limited to
loops with statically known iteration counts. Now, loops with dynamic operands
can be unrolled partially based on new thresholds. And this version offers:

                                             without                     with

loop-unrolling-variable-medium            7.3285+-0.0263     !      7.4737+-0.0300        ! definitely 1.0198x slower
loop-unrolling-variable                   0.3481+-0.0077            0.3477+-0.0082
loop-unrolling-variable-medium-dep       28.6089+-0.5330     ^     10.9441+-0.0331        ^ definitely 2.6141x faster
loop-unrolling-variable-large            66.6766+-0.0446     !     66.8410+-0.0559        ! definitely 1.0025x slower
loop-unrolling-variable-small             0.6056+-0.0088            0.6029+-0.0092

Currently, partial loop unrolling is disabled and it would be reenabled once
the heuristics tuning is done.

* JSTests/stress/loop-unrolling-partial.js: Added.
(func1):
(test):
* Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
(JSC::DFG::LoopUnrollingPhase::LoopData::isOperandConstant const):
(JSC::DFG::LoopUnrollingPhase::run):
(JSC::DFG::LoopUnrollingPhase::tryUnroll):
(JSC::DFG::LoopUnrollingPhase::identifyInductionVariable):
(JSC::DFG::LoopUnrollingPhase::shouldUnrollLoop):
(JSC::DFG::LoopUnrollingPhase::unrollLoop):
(JSC::DFG::LoopUnrollingPhase::LoopData::dump const):
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: <a href="https://commits.webkit.org/292691@main">https://commits.webkit.org/292691@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3b3c8afbfbd08bede506973e5d2e5f060fc3dcbe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96771 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16385 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6542 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101844 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47291 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16681 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24825 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73742 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30956 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99774 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12561 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87523 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54077 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12324 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5335 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46619 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/89444 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82418 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5419 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103867 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/95392 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23839 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17382 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82790 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24090 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83591 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82178 "Found 3 new API test failures: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/close-after-print (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20648 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26837 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4374 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17317 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23801 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28956 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/119019 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23460 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26940 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25201 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -473,30 +485,47 @@ class LoopUnrollingPhase : public Phase {
+        //
+        //  PreHeader                 PreHeader
+        //   |                          |
+        //  Header <-------        -> Header --------
```
Oops, this `Header` should be `HeaderBodyTailGraph_0`.

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
```diff
@@ -473,30 +485,47 @@ class LoopUnrollingPhase : public Phase {
+        //   |   |   | F  |        |    |           |
+        //   | T -----    | F      | T  | T         | F
+        //   |            |        |    |           |
+        //  BodyTailGraph -        -- BodyTailGraph |
```
And this `BodyTailGraph` should be `HeaderBodyTailGraph_1`.

justinmichaud reviewed (commented):
Nice. I wonder if we could detect some of these loops quickly before reaching FTL to emit a trip counter.

Constellation commented on Source/JavaScriptCore/runtime/OptionsList.h:
```diff
@@ -579,10 +579,13 @@ bool hasCapacityToUseLargeGigacage();
     v(Bool, dumpBaselineJITSizeStatistics, false, Normal, nullptr) \
     v(Bool, dumpDFGJITSizeStatistics, false, Normal, nullptr) \
     v(Bool, useLoopUnrolling, true, Normal, nullptr) \
+    v(Bool, enablePartialLoopUnrolling, false, Normal, nullptr) \
```
Let's rename it to `usePartialLoopUnrolling`

hyjorc1 committed:
[JSC] DFGLoopUnrollingPhase: Support partial loop unrolling for non-constant bounds
https://bugs.webkit.org/show_bug.cgi?id=290335
rdar://147775029

Reviewed by Yusuke Suzuki.

This patch enhances the DFG loop unrolling phase to support partial unrolling
for loops with non-constant bounds. Previously, loop unrolling was limited to
loops with statically known iteration counts. Now, loops with dynamic operands
can be unrolled partially based on new thresholds. And this version offers:

                                             without                     with

loop-unrolling-variable-medium            7.3285+-0.0263     !      7.4737+-0.0300        ! definitely 1.0198x slower
loop-unrolling-variable                   0.3481+-0.0077            0.3477+-0.0082
loop-unrolling-variable-medium-dep       28.6089+-0.5330     ^     10.9441+-0.0331        ^ definitely 2.6141x faster
loop-unrolling-variable-large            66.6766+-0.0446     !     66.8410+-0.0559        ! definitely 1.0025x slower
loop-unrolling-variable-small             0.6056+-0.0088            0.6029+-0.0092

Currently, partial loop unrolling is disabled and it would be reenabled once
the heuristics tuning is done.

* JSTests/stress/loop-unrolling-partial.js: Added.
(func1):
(test):
* Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
(JSC::DFG::LoopUnrollingPhase::LoopData::isOperandConstant const):
(JSC::DFG::LoopUnrollingPhase::run):
(JSC::DFG::LoopUnrollingPhase::tryUnroll):
(JSC::DFG::LoopUnrollingPhase::identifyInductionVariable):
(JSC::DFG::LoopUnrollingPhase::shouldUnrollLoop):
(JSC::DFG::LoopUnrollingPhase::unrollLoop):
(JSC::DFG::LoopUnrollingPhase::LoopData::dump const):
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: https://commits.webkit.org/292691@main
